docs/development/design/network-policy.md

```diff
@@ -0,0 +1,256 @@
+- A device can potentially have multiple policy groups applied at the same time.
+
+## Default Security Group Rules
+
+The default security group rules will be an explicit allow, adopting the EC2 approach of an implicit deny. The agent will add an explicit deny at the end of the ingress or egress chain. The user only sees an implicit deny with explicit allows just as in Ec2.
+
+- Ordering will be done by the order the user installs the rules. This is possible since there are no denies. As a refernce, you can compare EC2 rules to Azure rules for not allowing deny statements vs allowing deny statements. Order begins to matter when denies are in place. This adds complexity which for our use case does not add any clear value.
+- Users can add ranges of a given field. For example both, IpRanges with a value of `100.100.0.100-100.100.0.120` is valid and a prefix such as `100.100.0.128/25` is also valid. Along with that, a single address such as `100.100.0.10`.
+- The same applies to source port and destination ports, `PortFrom:8080` coupled with `PortTo:9000` would equate to a rule of `8080-9000` being permitted. `PortFrom:0 PortTo:0` will be read as `ip permit <protocol> any`. `PortFrom:443 PortTo:443` would be equivalent to `ip permit <protocol> 4434`.
+- L3 `IpRanges` are applied based on the direction field they are located in SecurityGroups. `InboundRules` have the IP prefix applied to the `saddr` field in nftables in the input chain, while `OutboundRules` are applied to the `daddr` field in the outbound chain.
+- In regard to L4 ports, destination ports are what will be supported. For example, an ingress rule of `input tcp dport 22 counter accept` would apply port 22 to the L4 dport value in the nexodus-inbound chain meaning any host can connect to the node on port 22.
+
+## Security Group User Interface
+
+- The user can add rules via web UI, the nexctl tool or the http API. There would not be support for adding or manipulating rules via the agent.
+
+## Rule Deconfliction
+
+- The deconfliction of user provided rules is managed by nft. Let's look at the following example where a user defines a permit `icmp6 any` and an `icmp6 2001:0db8:1337:cafe::/64`. The JSON would look as follows:
+
+```json
+  {
+    "ip_protocol": "icmp",
+    "from_port": 0,
+    "to_port": 0,
+    "ip_ranges": [
+      "2001:0db8:1337:cafe::/64"
+    ]
+  },
+  {
+    "ip_protocol": "icmp",
+    "from_port": 0,
+    "to_port": 0
+  }
+```
+
+The actual rule in nftables would only insert the LPM (Longest Prefix Match). The resulting chain would look as follows.
+
+```plaintext
+table inet nexodus {
+	chain nexodus-inbound {
+		type filter hook input priority filter; policy accept;
+		iifname "wg0" ct state established,related counter packets 0 bytes 0 accept
+	}
+
+	chain nexodus-outbound {
+		type filter hook output priority filter; policy accept;
+	}
+```
+
+The same LPM rule optimizations also applies to IPv4 and IPv6.
+
+## Default Security Group
```

Comment:
Perhaps merge this with the previous section about default security group rules? I was wondering what the default was when reading that section.

Reply:
Oops, I duped that. Pretty much wrote the same section twice 🤦 The more detail there the better though so not a total waste.